pkg/containers/containers.go

```diff
@@ -242,6 +202,52 @@ func (c *Containers) CgroupRemove(cgroupId uint64) {
 	c.deleted = append(c.deleted, cgroupId)
 }
 
+func (c *Containers) CgroupMkdir(cgroupId uint64, subPath string) (CgroupInfo, error) {
+	curTime := time.Now()
+	// first try to find this new cgroup in cgroup fs
+	path, err := getCgroupPath(c.cgroupMP, cgroupId, subPath)
+	if err == nil {
+		// cgroup entry found in cgroupfs
+		var stat syscall.Stat_t
+		if err := syscall.Stat(path, &stat); err == nil {
+			return c.CgroupUpdate(cgroupId, path, time.Unix(stat.Ctim.Sec, stat.Ctim.Nsec))
+		}
+	}
+
+	// if no entry found, container may have already exited
+	// update map with existing data (ctime becomes an estimation only)
+	return c.CgroupUpdate(cgroupId, subPath, curTime)
+}
+
+func getCgroupPath(rootDir string, cgroupId uint64, subPath string) (string, error) {
```

Comment:
This function is ingenious but I think it should be split into the 2 cases it represents. It looks overcomplicated the way it is.

1. **If subPath string is empty**, then it recurses the cgroup root dir (and subPath should be "" in the recursive call, because subPath will always be related to the cgroup fs root directory). Example:

```
subPath:
entryPath: /sys/fs/cgroup/system.slice/console-setup.service
subPath:
entryPath: /sys/fs/cgroup/system.slice/containerd.service
subPath:
entryPath: /sys/fs/cgroup/system.slice/cpufrequtils.service
subPath:
entryPath: /sys/fs/cgroup/system.slice/cron.service
subPath:
entryPath: /sys/fs/cgroup/system.slice/dbus.service
subPath:
entryPath: /sys/fs/cgroup/system.slice/dbus.socket
subPath:
entryPath: /sys/fs/cgroup/system.slice/dm-event.socket
subPath:
entryPath: /sys/fs/cgroup/system.slice/docker-4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec.scope
subPath:
```

2. **If subPath isn't empty**, then there is no recursion and subPath is related to the cgroup fs root dir:

```
cgroupId = 33264, path = /sys/fs/cgroup/system.slice/docker-4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec.scope, ContainerId = 4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec, Runtime = docker

entryPath: /sys/fs/cgroup/system.slice/docker-4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec.scope
subPath: /system.slice/docker-4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec.scope

PATH: /sys/fs/cgroup/system.slice/docker-4bed3dee01485506ad82b5c2d9d66d9daf58f94e8d982e7b6f080457a09a1dec.scope
```

Reply:
Why do you say that there will be no recursion if subPath is not empty? There will be a recursion in both cases. The point is to find a path in cgroupfs that has subPath as a suffix. If subPath is empty, we will just look for cgroupid (walking cgroupfs recursively), and if subPath is not empty, we will still recurse, but will save stating to irrelevant paths that don't have subPath as a suffix